runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java

```diff
@@ -59,33 +59,19 @@ public PipelineResult run(Pipeline pipeline) {
     } catch (Throwable e) {
       // Special case hack to pull out assertion errors from PAssert; instead there should
       // probably be a better story along the lines of UserCodeException.
-      Throwable cause = e;
-      Throwable oldCause = e;
-      PipelineExecutionException executionException = null;
-      do {
-
-        // find UserCodeException and throw PipelineExecutionException
-        if (cause instanceof UserCodeException) {
-          executionException = new PipelineExecutionException(cause.getCause());
-        }
-
-        if (cause.getCause() == null) {
-          break;
-        }
-
-        oldCause = cause;
-        cause = cause.getCause();
-
-      } while (!oldCause.equals(cause));
-      if (cause instanceof AssertionError) {
-        throw (AssertionError) cause;
-      } else {
-        if (executionException != null) {
-          throw executionException;
-        } else {
-          throw e;
+      UserCodeException innermostUserCodeException = null;
+      for (; e.getCause() != null; e = e.getCause()) {
```

Comment:
I'm pretty opposed to assigning over the caught `Throwable`. Reassigning to a local variable is preferable.